homeassistant/components/openrgb/__init__.py

```diff
@@ -0,0 +1,213 @@
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_HOST): cv.string,
+                    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+                    vol.Optional(CONF_CLIENT_ID, default=DEFAULT_CLIENT_ID): cv.string,
+                }
+            )
+        },
+    ),
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass, config):
+    """Set up the OpenRGB integration."""
+    conf = config.get(DOMAIN)
+    if conf is not None:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up OpenRGB platform."""
+    try:
+        orgb = OpenRGBClient(
+            entry.data[CONF_HOST],
+            entry.data[CONF_PORT],
+            name=entry.data[CONF_CLIENT_ID],
+        )
+    except ConnectionError as err:
+        _LOGGER.error("Connection error during integration setup. Error: %s", err)
+        return False
+
+    def connection_recovered():
+        if not hass.data[DOMAIN]["online"]:
+            _LOGGER.info(
+                "Connection reestablished to OpenRGB SDK Server at %s:%i",
+                entry.data[CONF_HOST],
+                entry.data[CONF_PORT],
+            )
+
+        hass.data[DOMAIN]["online"] = True
+        async_dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+
+    def connection_failed():
+        if hass.data[DOMAIN]["online"]:
+            hass.data[DOMAIN][ORGB_DATA].comms.stop_connection()
+            _LOGGER.info(
+                "Connection lost to OpenRGB SDK Server at %s:%i",
+                entry.data[CONF_HOST],
+                entry.data[CONF_PORT],
+            )
+
+        hass.data[DOMAIN]["online"] = False
+        async_dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+
+    hass.data[DOMAIN] = {
+        "online": True,
+        ORGB_DATA: orgb,
+        ORGB_TRACKER: None,
+        ENTRY_IS_SETUP: set(),
+        "entities": {},
+        "pending": {},
+        "connection_failed": connection_failed,
+        "connection_recovered": connection_recovered,
+    }
+
+    # Initial device load
+    async def async_load_devices(device_list):
+        device_type_list = {}
+
+        for device in device_list:
+            ha_type = "light"
+            if ha_type not in device_type_list:
+                device_type_list[ha_type] = []
+            device_type_list[ha_type].append(device)
+
+            entity_id = orgb_entity_id(device)
+            if entity_id not in hass.data[DOMAIN]["entities"]:
+                hass.data[DOMAIN]["entities"][entity_id] = None
+
+        for ha_type, dev_ids in device_type_list.items():
+            config_entries_key = f"{ha_type}.openrgb"
+
+            if config_entries_key not in hass.data[DOMAIN][ENTRY_IS_SETUP]:
```

Comment:
The purpose of this loop took me awhile to get my head around. I think it'd be more clear if this was called PLATFORM_IS_SETUP since what it is tracking is if the config entity setup has been forwarded to the platform. Entity is a very overloaded word in the HA universe unfortunately.

Reply:
You might also consider always setting up all platforms (currently just light) so there is no need to track this and the async_setup_entry function always calls forward and pull devices service implementation always calls the discover signal. It might reduce complexity of this function's design significantly.